src/userauth.c

```diff
@@ -118,21 +121,62 @@ static char *userauth_list(LIBSSH2_SESSION *session, const char *username,
     }
 
     if(session->userauth_list_state == libssh2_NB_state_sent) {
-        rc = _libssh2_packet_requirev(session, reply_codes,
-                                      &session->userauth_list_data,
-                                      &session->userauth_list_data_len, 0,
-                                      NULL, 0,
-                               &session->userauth_list_packet_requirev_state);
-        if(rc == LIBSSH2_ERROR_EAGAIN) {
-            _libssh2_error(session, LIBSSH2_ERROR_EAGAIN,
-                           "Would block requesting userauth list");
-            return NULL;
-        }
-        else if(rc || (session->userauth_list_data_len < 1)) {
-            _libssh2_error(session, rc, "Failed getting response");
-            session->userauth_list_state = libssh2_NB_state_idle;
-            return NULL;
-        }
+        do {
+            state = &session->userauth_list_packet_requirev_state;
+            rc = _libssh2_packet_requirev(session, reply_codes,
+                                          &session->userauth_list_data,
+                                          &session->userauth_list_data_len, 0,
+                                          NULL, 0, state);
+            if(rc == LIBSSH2_ERROR_EAGAIN) {
+                _libssh2_error(session, LIBSSH2_ERROR_EAGAIN,
+                               "Would block requesting userauth list");
+                return NULL;
+            }
+            else if(rc || (session->userauth_list_data_len < 1)) {
+                _libssh2_error(session, rc, "Failed getting response");
+                session->userauth_list_state = libssh2_NB_state_idle;
+                return NULL;
+            }
+            if(session->userauth_list_data[0] == SSH_MSG_USERAUTH_BANNER &&
+               (session->userauth_list_data_len >= 5)) {
+
+                methods_len =
+                  _libssh2_ntohu32(session->userauth_list_data + 1);
+
+                /* Cap to 512 bytes. */
```

Comment:
The define is 2048, it appears this comment needs updating.

Reply:
Thanks for pointing this out. It should be fixed now.